### ISSUE TITLE
Python code for TEC prediction LSTM

### ISSUE THREAD

**Issue by S9ra** [S9ra/Code-for-LSTM#1]
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential, load_model
from tensorflow.keras.layers import LSTM, Dense, Dropout
import matplotlib.pyplot as plt
from google.colab import drive
drive.mount('/content/drive')

url = 'https://drive.google.com/uc?export=download&id=16C910WW4cCKUysDlLyD1swM19nt1Fbva'
data = pd.read_csv(url)
features = data[['Year', 'DateOfYear', 'SunspotNumber', 'ApIndex', 'IG12Index', 'LowEIonElectronFlux', 'F107', 'Latitude', 'Longitude', 'Altitude']]
target = data['TEC']
scaler_features = MinMaxScaler()
scaled_features = scaler_features.fit_transform(features)
scaler_tec = MinMaxScaler()
scaled_target = scaler_tec.fit_transform(target.values.reshape(-1, 1))

def create_lstm_data(features, target, time_steps=10):
    X, y = [], []
    for i in range(len(features) - time_steps):
        X.append(features[i:i + time_steps])
        y.append(target[i + time_steps])
    return np.array(X), np.array(y)

time_steps = 10
X, y = create_lstm_data(scaled_features, scaled_target, time_steps=time_steps)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=False)

model = Sequential()
model.add(LSTM(200, return_sequences=True, input_shape=(X_train.shape[1], X_train.shape[2])))
model.add(Dropout(0.3))
model.add(LSTM(200, return_sequences=False))
model.add(Dropout(0.3))
model.add(Dense(50))
model.add(Dense(1))

# Compile the model using a different optimizer
model.compile(optimizer='rmsprop', loss='mean_squared_error')

# Train the model (increased epochs)
model.fit(X_train, y_train, batch_size=32, epochs=200)  # Increased epochs

# 1. Save the Entire Model (Architecture + Weights + Optimizer State)
model.save('/content/drive/My Drive/my_lstm_model.h5') # Save the model

# Make predictions on the test set
y_pred = model.predict(X_test)

# Inverse transform the predicted and actual values to original scale
y_pred_original = scaler_tec.inverse_transform(y_pred)
y_test_original = scaler_tec.inverse_transform(y_test)

# Plot Original, Actual, and Predicted TEC
plt.figure(figsize=(12, 6))

# Plot original TEC (entire dataset)
plt.plot(target, label='Original TEC', color='green', alpha=0.5)

# Plot actual TEC for the training set
plt.plot(range(len(y_train)), scaler_tec.inverse_transform(y_train), label='Training Set Actual TEC', color='blue', alpha=0.7)

# Plot actual TEC for the test set
plt.plot(range(len(y_train), len(y_train) + len(y_test_original)), y_test_original, label='Test Set Actual TEC', color='orange')

# Plot predicted TEC for the test set
plt.plot(range(len(y_train), len(y_train) + len(y_pred_original)), y_pred_original, label='Predicted TEC', color='red', linestyle='dashed')

plt.title('Original vs Actual vs Predicted TEC')
plt.xlabel('Time Steps')
plt.ylabel('TEC')
plt.legend()
plt.show()
